### ISSUE TITLE
Postgres connection error

### ISSUE THREAD

**Issue by TimRepke** [ambiverse-nlu/ambiverse-nlu#5]
I'm trying to set up an Ambiverse REST endpoint on one of our servers so we can query it locally. So far the setup is rather easy, the Postgres database is running with the imported AIDA dump and the maven build succeeded. 

However, when starting the web_service.sh, ambiverse fails to establish a connection ( [it logs timeout warnings](https://github.com/ambiverse-nlu/ambiverse-nlu/blob/341dc62abf47396aa04e6d0ca7c9e071b5d5b06e/src/main/java/de/mpg/mpi_inf/ambiversenlu/nlu/entitylinking/EntityLinkingManager.java#L206) ).
```
2019-03-18 16:17:01,577 [main] INFO nlu.entitylinking.EntityLinkingManager:206  - Postgres DB seems unavailable. This is expected during first startup using Docker. Waiting for 60s in addition (already waited 0s, will wait up to 10800s in total).
```

To figure out what's going on, I checked whether all config parameters are used by the code correctly and also added `prop.put("dataSource.logWriter", new PrintWriter(System.out));` to get a more verbose log of the database connection library. Everything seems fine, however `loginTimeout=0` and `socketTimeout=0` seems odd. Changing it to a non-zero value doesn't change the behaviour, where the library quickly logs a lot of connection attempts.
To check if anything gets to the postgres server, I configured it to log everything:
```
2019-03-18 16:16:52.753 CET [5573] [unknown]@[unknown] LOG:  00000: connection received: host=127.0.0.1 port=40314
2019-03-18 16:16:52.753 CET [5573] [unknown]@[unknown] LOCATION:  BackendInitialize, postmaster.c:4205
2019-03-18 16:16:52.754 CET [5573] ambiverse@aida_20180120_cd_de_en_es_ru_zh_v18 LOG:  00000: connection authorized: user=ambiverse database=aida_20180120_cd_de_en_es_ru_zh_v18
2019-03-18 16:16:52.754 CET [5573] ambiverse@aida_20180120_cd_de_en_es_ru_zh_v18 LOCATION:  PerformAuthentication, postinit.c:279
2019-03-18 16:16:52.757 CET [5573] ambiverse@aida_20180120_cd_de_en_es_ru_zh_v18 LOG:  00000: disconnection: session time: 0:00:00.004 user=ambiverse database=aida_20180120_cd_de_en_es_ru_zh_v18 host=127.0.0.1 port=40314
2019-03-18 16:16:52.757 CET [5573] ambiverse@aida_20180120_cd_de_en_es_ru_zh_v18 LOCATION:  log_disconnections, postgres.c:4614
2019-03-18 16:16:53.760 CET [5593] [unknown]@[unknown] LOG:  00000: connection received: host=127.0.0.1 port=40316
2019-03-18 16:16:53.760 CET [5593] [unknown]@[unknown] LOCATION:  BackendInitialize, postmaster.c:4205
2019-03-18 16:16:53.762 CET [5593] ambiverse@aida_20180120_cd_de_en_es_ru_zh_v18 LOG:  00000: connection authorized: user=ambiverse database=aida_20180120_cd_de_en_es_ru_zh_v18
2019-03-18 16:16:53.762 CET [5593] ambiverse@aida_20180120_cd_de_en_es_ru_zh_v18 LOCATION:  PerformAuthentication, postinit.c:279
2019-03-18 16:16:53.767 CET [5593] ambiverse@aida_20180120_cd_de_en_es_ru_zh_v18 LOG:  00000: disconnection: session time: 0:00:00.006 user=ambiverse database=aida_20180120_cd_de_en_es_ru_zh_v18 host=127.0.0.1 port=40316
2019-03-18 16:16:53.767 CET [5593] ambiverse@aida_20180120_cd_de_en_es_ru_zh_v18 LOCATION:  log_disconnections, postgres.c:4614
2019-03-18 16:16:54.770 CET [5635] [unknown]@[unknown] LOG:  00000: connection received: host=127.0.0.1 port=40318
2019-03-18 16:16:54.770 CET [5635] [unknown]@[unknown] LOCATION:  BackendInitialize, postmaster.c:4205
2019-03-18 16:16:54.772 CET [5635] ambiverse@aida_20180120_cd_de_en_es_ru_zh_v18 LOG:  00000: connection authorized: user=ambiverse database=aida_20180120_cd_de_en_es_ru_zh_v18
2019-03-18 16:16:54.772 CET [5635] ambiverse@aida_20180120_cd_de_en_es_ru_zh_v18 LOCATION:  PerformAuthentication, postinit.c:279
2019-03-18 16:16:54.777 CET [5635] ambiverse@aida_20180120_cd_de_en_es_ru_zh_v18 LOG:  00000: disconnection: session time: 0:00:00.006 user=ambiverse database=aida_20180120_cd_de_en_es_ru_zh_v18 host=127.0.0.1 port=40318
2019-03-18 16:16:54.777 CET [5635] ambiverse@aida_20180120_cd_de_en_es_ru_zh_v18 LOCATION:  log_disconnections, postgres.c:4614
```
(yes, my user isn't ambiversenlu, it's just ambiverse).   
Connecting to the server from pgadmin looks exactly the same in the logs, except that it doesn't immediately disconnect again.

I have tried a lot, but I guess it boils down to the scenario described above, where the database connection isn't staying alive. I have tried the git version tagged 1.0 and 1.1 as well as current master.

```
% cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic

% psql --version                                                                                                                                                                          
psql (PostgreSQL) 11.2 (Ubuntu 11.2-1.pgdg18.04+1)

% mvn -version
Apache Maven 3.5.2
Maven home: /usr/share/maven
Java version: 1.8.0_201, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-8-oracle/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "4.15.0-32-generic", arch: "amd64", family: "unix"
```

**Comment by hoffart** [comment]
Could it be a typo? The db is called aida_20180120_cs_de_en_es_ru_zh_v18 (note the **cs** not **cd**)

**Comment by TimRepke** [comment]
@hoffart thanks for the sharp eye. unfortunately the typo already happened when I created the database. so in this case it is spelled correctly.

**Comment by dmilcevski** [comment]
One thing that I can think of is that maybe the driver version we use in the .pom is not compatible with the server version. We use postgresql 9.4 and I see you are using 11.2. Try changing the driver version (not sure what other changes might this require) or downgrade the server version (if you are using docker).

**Comment by malteos** [comment]
When you use the command from the README, the Postgres container does not start correctly, because it cannot download the dump files:

```
docker run -d --name nlu-db-postgres \
  -p 5432:5432 \
  -e POSTGRES_DB=aida_20180120_b3_de_en_v18_db \
  -e POSTGRES_USER=ambiversenlu \
  -e POSTGRES_PASSWORD=ambiversenlu \
  ambiverse/nlu-db-postgres
```

The container trys to do download the dump from: 
- `http://ambiversenlu-download.mpi-inf.mpg.de/postgres/aida_20180120_b3_de_en_v18_db.sql.gz` 
- correct is (with out `_db`): http://ambiversenlu-download.mpi-inf.mpg.de/postgres/aida_20180120_b3_de_en_v18.sql.gz

I fixed this by changing the Postgres database name to `aida_20180120_b3_de_en_v18`:

```
docker run -d --name nlu-db-postgres \
  -p 5432:5432 \
  -e POSTGRES_DB=aida_20180120_b3_de_en_v18 \
  -e POSTGRES_USER=ambiversenlu \
  -e POSTGRES_PASSWORD=ambiversenlu \
  ambiverse/nlu-db-postgres
```

Also, I changed the database hostname `127.0.0.1` in the corresponding config file `src/main/config/aida_20180120_b3_de_en_v18_db/database_aida.properties`. Sometimes containers are not reachable with `localhost`.

**Comment by KartikTaskhuman** [comment]
Using the database name aida_20180120_b3_de_en_v18 works as mentioned by @malteos but for the database aida_20180120_cs_de_en_es_ru_zh_v18 I still get the postgres connection error. 

> 2019-04-29 19:13:17,642 [main] WARN  hadoop.util.NativeCodeLoader:62  - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
> 2019-04-29 19:13:17,981 [main] INFO  internal.logging.Slf4JLogger:101  - Your platform does not provide complete low-level API for accessing direct buffers reliably. Unless explicitly requested, heap buffer will always be preferred to avoid potential system unstability.
> 2019-04-29 19:13:18,620 [sparkDriverActorSystem-akka.actor.default-dispatcher-3] INFO  event.slf4j.Slf4jLogger$$anonfun$receive$1$$anonfun$applyOrElse$3:74  - Starting remoting
> 2019-04-29 19:13:18,859 [sparkDriverActorSystem-akka.actor.default-dispatcher-3] INFO  event.slf4j.Slf4jLogger$$anonfun$receive$1$$anonfun$applyOrElse$3:74  - Remoting started; listening on addresses :[akka.tcp://sparkDriverActorSystem@10.64.13.31:33687]
> 2019-04-29 19:13:23,651 [sparkDriverActorSystem-akka.actor.default-dispatcher-4] INFO  event.slf4j.Slf4jLogger$$anonfun$receive$1$$anonfun$applyOrElse$3:74  - Shutting down remote daemon.
> 2019-04-29 19:13:23,653 [sparkDriverActorSystem-akka.actor.default-dispatcher-4] INFO  event.slf4j.Slf4jLogger$$anonfun$receive$1$$anonfun$applyOrElse$3:74  - Remote daemon shut down; proceeding with flushing remote transports.
> 2019-04-29 19:13:23,764 [sparkDriverActorSystem-akka.actor.default-dispatcher-4] INFO  event.slf4j.Slf4jLogger$$anonfun$receive$1$$anonfun$applyOrElse$3:74  - Remoting shut down.
> 2019-04-29 19:13:58,194 [main] INFO  nlu.entitylinking.EntityLinkingManager:206  - Postgres DB seems unavailable. This is expected during first startup using Docker. Waiting for 60s in addition (already waited 0s, will wait up to 10800s in total).
> 2019-04-29 19:15:28,196 [main] INFO  nlu.entitylinking.EntityLinkingManager:206  - Postgres DB seems unavailable. This is expected during first startup using Docker. Waiting for 60s in addition (already waited 60s, will wait up to 10800s in total).
> 2019-04-29 19:16:58,198 [main] INFO  nlu.entitylinking.EntityLinkingManager:206  - Postgres DB seems unavailable. This is expected during first startup using Docker. Waiting for 60s in addition (already waited 120s, will wait up to 10800s in total).

aida_20180120_cs_de_en_es_ru_zh_v18 is much larger database than aida_20180120_b3_de_en_v18.
I am able to connect manually through psql manually using the correct database credentials and server hosts (both as 127.0.0.1, 0.0.0.0) but it fails when I run the start_webservice.sh

Could someone help triage this? It seems like start_webservice.sh is broken when we use a larger database specified above, but it works for all other smaller configurations. As I see in the following issue: https://github.com/ambiverse-nlu/ambiverse-nlu/issues/4 only way to replicate the behaviour at https://ambiversenlu.mpi-inf.mpg.de/ is to use the aida_20180120_cs_de_en_es_ru_zh_v18. FYI I am unable to simulate the behaviour in the above demo page by using the small database dump at aida_20180120_b3_de_en_v18 (I got some entities missing when I compared to the demo)

Any thoughts?

**Comment by hoffart** [comment]
The demo is using the full database, so results are not replicable using the small database.

How long did you wait in total for the database? It is downloading/unpacking/importing/indexing a 100GB data file, this can take a couple of hours.

**Comment by TimRepke** [comment]
I downloaded the database dump and imported it to Postgres. So the import and db startup is already done before starting the ambiverse scripts. Importing did take quite a while but I didn't stop the time. I started it on the server and let it run over night.

Changing the driver version doesn't fully work straight away unfortunately - I'd have to fix some deprecation. Once I have time to look into this again, I'll make sure to push the updated sources if I manage to fix it.

**Comment by dylanhouxinglin** [comment]
I meet the same problem with you, have u got any solution for this issues now?

**Comment by TimRepke** [comment]
Hi @dylanhouxinglin ,
actually, I did not. I kept on trying for a while, gave up, and used their demo server by spreading requests over a few days.

**Comment by ghost** [comment]
What is the request looks like?
